dali/kernels/common/join/tensor_join_cpu.h

```diff
@@ -0,0 +1,155 @@
+  TensorShape<> ret;
+  auto &insh = in_shapes[0];
+  ret.resize(insh.size() + 1);
+  for (int i = 0; i < axis; i++) {
+    ret[i] = insh[i];
+  }
+  ret[axis] = in_shapes.size();
+  for (int i = axis + 1; i < ret.size(); i++) {
+    ret[i] = insh[i - 1];
+  }
+  return ret;
+}
+
+
+template<>
+TensorShape<> DetermineShape<CONCAT>(span<const TensorShape<>> in_shapes, int axis) {
+  TensorShape<> ret = in_shapes[0];
+  ret[axis] *= in_shapes.size();
+  return ret;
+}
+
+}  // namespace detail
+
+template<typename Out, typename In, JoinMode mode = STACK, int dims = -1>
+struct TensorJoinCpu {
+  KernelRequirements Setup(KernelContext &ctx, span<const TensorShape<dims>> in_shapes, int axis) {
+    n_input_tensors_ = in_shapes.size();
+    orig_shapes_(in_shapes);
+    auto ndims = in_shapes[0].sample_dim();
+    DALI_ENFORCE(axis < ndims && axis > -ndims, "Incorrect axis. Actual: ", axis, ". Expected in [",
+                 -ndims + 1, ", ", ndims - 1, "] interval");
+    axis_ = axis >= 0 ? axis : ndims + axis;
+
+    {
+      const auto &ref = in_shapes[0];
+      for (int i = 1; i < n_input_tensors_; i++) {
+        DALI_ENFORCE(in_shapes[i].sample_dim() == ref.sample_dim(),
+                     "Every input shape must have the same number of dimensions.");
+        for (int j = 0; j < ref.size(); j++) {
+          if (mode == CONCAT) {
+            DALI_ENFORCE(in_shapes[i][j] == ref.shape[j] || (j == axis_ && mode == CONCAT),
+                         make_string(
+                                 "Number of samples in every dimension "
+                                 "(but the one along which concatenation occurs) must be the same "
+                                 "(CONCAT mode). 0-th shape at index ", j, " has dimension ",
+                                 ref.shape[j], ", while ", i, "-th shape at index ", j,
+                                 " has dimension ", in_shapes[i][j]));
+          } else {
+            DALI_ENFORCE(in_shapes[i][j] == ref.shape[j], make_string(
+                    "Number of samples in every dimension must be the same (STACK mode). "
+                    "0-th shape at index ", j, " has dimension ", ref.shape[j], ", while ", i,
+                    "-th shape at index ", j, " has dimension ", in_shapes[i][j]));
+          }
+        }
+      }
+    }
+
+    KernelRequirements kr;
+    output_shape_ = detail::DetermineShape<mode>(in_shapes, axis);
+    kr.output_shapes.resize(1);
+    TensorListShape<> tmp({output_shape_});  // clang's destructor bug still haunting
+    kr.output_shapes[0] = tmp;
+    return kr;
+  }
+
+
+  KernelRequirements Setup(KernelContext &ctx, span<const InTensorCPU<In, dims>> in, int axis) {
+    std::vector<TensorShape<>> in_shapes(in.size());
+    for (int i = 0; i < in.size(); i++) {
+      in_shapes[i] = in[i].shape;
+    }
+    return Setup(ctx, in_shapes, axis);
+  }
+
+
+  void Run(KernelContext &ctx, const OutTensorCPU<Out, dims> &out,
```

Comment:
When stacking, output will have an extra dimension.
```suggestion
  static constexpr int output_dims = (dims == DynamicDimensions ? DynamicDimensions : mode == STACK ? dims + 1 : dims);
  void Run(KernelContext &ctx, const OutTensorCPU<Out, output_dims> &out,
```